Update to implements capacitor v3

rdlabo opened the pull request:
This pull request is update to implements capacitor v3
This is branch of https://github.com/capacitor-community/admob/tree/feat/v3 and test here:

```
npm i --save @capacitor-community/admob@next
```

and you can confirm README: https://github.com/capacitor-community/admob/blob/feat/v3/README.md

# iOS
- [x] Adaptive Banner
- [x] Interstitial to `GADInterstitialAd `
- [x] Reward to `GADRewardedAd`

https://developers.google.com/admob/ios/ios14

# Android
- [x] Adaptive Banner
- [x] Interstitial to `InterstitialAd `
- [x] Reward to `RewardedAd`
- [x] Tests

# Common
- [x] All Work Tests

https://developers.google.com/admob/android/migration

rdlabo commented:
`com.google.android.gms:play-services-ads:20.0.0` is released:
https://developers.google.com/android/guides/releases

rdlabo commented:
@distante Hi. I worked this plugin updated to implements capacitor v3. But Android test is hard work for me. Can you help this?

And I want your review at code of Adaptive Banner:
https://github.com/capacitor-community/admob/blob/feat/v3/android/src/main/java/com/getcapacitor/community/admob/executors/BannerExecutor.java#L66-L69

Thanks.

distante commented:
Will do as soon as possible.

distante commented:
I am having problems running the demo project @rdlabo, do you have some special workflow for it?

I always get write permission problem when I try to install the angular demo project dependencies

distante commented:
> I am having problems running the demo project @rdlabo, do you have some special workflow for it?
> 
> I always get write permission problem when I try to install the angular demo project dependencies

For the time being I declared the demo app dependencies hard in package.json and it works now. (besides AdMob).

distante commented on android/src/main/java/com/getcapacitor/community/admob/executors/AdRewardExecutor.java:
```diff
@@ -29,83 +32,22 @@ public AdRewardExecutor(
                     () -> {
                         final AdRequest adRequest = RequestHelper.createRequest(adOptions);
                         final String id = AdViewIdHelper.getFinalAdId(adOptions, adRequest, logTag, contextSupplier.get());
-                        mRewardedVideoAd.loadAd(id, adRequest);
-                        mRewardedVideoAd.setRewardedVideoAdListener(getRewardedVideoAdListener(call, notifyListenersFunction));
```
@rdlabo I am having troubles understanding the API change for `RewardAds`. The notifications for successfull Reward item is removed from `definitions.ts` here and the implementation is also removed here from the Executor. 

https://github.com/capacitor-community/admob/commit/93182951c96dee8df4cc288ea80466391e49192b#diff-f1f3c6677661c21128a30f64358c6615c6801043e1cbfcc904302422a17b3246L112-L115

That is the reason why the AdRewardExecutor test is failing here:

![image](https://user-images.githubusercontent.com/5490201/114300302-ebc63a80-9abf-11eb-861c-ae2f7dd3a26c.png)

Why are those events removed?

rdlabo replied:
@distante New Reward API can get item from `showRewardVideoAd(item)`.
https://github.com/capacitor-community/admob/commit/93182951c96dee8df4cc288ea80466391e49192b#diff-f1f3c6677661c21128a30f64358c6615c6801043e1cbfcc904302422a17b3246R78

And AdMob API is changed to be FullScreen Event. So I removed this api.

distante replied:
But I think  the change in the implementation of the  AdMob library should not be directly attach to the Plugin API.

The capacitor users still need to know that a reward was given and proxy events like "rewarded, failed", etc exist exist. 

Currently the only way is to wait for a promise call, but for angular users, events are more Rxjs Observables friendly.

This would meant a really big change for the Capacitor users.

I think we should keep those events since it is just a couple lines of code.

rdlabo replied:
> events are more Rxjs Observables friendly.

I forgot this point of view. Thanks and agree!!
But I suggest change to API name `onUserEarnedReward` .  `onAdLoaded` is a misleading name.

distante replied:
I think I will enable Kotlin support so I can create clean Kotlin enums with all the events names

rdlabo commented:
@distante Sorry for late. special workflow is none for my environment🤔(MacOS)

1. npm install
2. npm run build
3. cd demo/angular && npm run install && npm run cap

But I'm not particular about this, so change `dependencies hard` looks good.

distante commented:
@rdlabo take a look at the last commit.

I am not done with all the events but to not get mixed or forgotten events I have separate them into sub definitions and grouping them together in `src/definitions.ts`

```
type AdMobDefinitions = BannerDefinitions & RewardDefinitions & InterstitialDefinitions;

export interface AdMobPlugin extends AdMobDefinitions {
   // general admon plugin calls
}
```

I also created an enum for the reward events
```
export enum RewardAdPluginEvents {
  Loaded= 'onRewardedVideoAdLoaded',
  FailedToLoad= 'onRewardedVideoAdFailedToLoad',
  Showed= 'onRewardedVideoAdShowed',
  FailedToShow= 'onRewardedVideoAdFailedToShow',
  Dismissed= 'onRewardedVideoAdDismissed',
  Rewarded= 'onRewardedVideoAdReward',
}
```

in typescript, the plugin consumer has to write :

```
Admob.addEventListener(RewardAdPluginEvents.Load, () => {
});
```

So there is no change that the user write a wrong string, the value of each enum (like `onRewardedVideoAdLoaded`)  is what I am using on the Android part to also emit events:

```
package com.getcapacitor.community.admob.models

enum class RewardAdPluginEvents(val webEventName: String) {
    Loaded("onRewardedVideoAdLoaded"),
    FailedToLoad("onRewardedVideoAdFailedToLoad"),
    Showed("onRewardedVideoAdShowed"),
    FailedToShow("onRewardedVideoAdFailedToShow"),
    Dismissed("onRewardedVideoAdDismissed"),
    Rewarded("onRewardedVideoAdReward")
}
```

Now, I am not sure how to work with the `Promises` in our Use Cases.

For example a typescript `AdMob.prepareRewardVideoAd()` can have two branches in the native part:
1. Happy Path: Load Successful
2. Sad Path: Load unsuccessful.

We can define the type of Happy Path with:

```
prepareRewardVideoAd(options: AdOptions): Promise<{adUnitId: string}>;
```

But Typescript does not support Error definitions for Promises. They are always `any`

`AdMob.showRewardVideoAd()` is harder because It can have 3
1. Happy Path: The video is Showed.
2. Happy Path: The video is Showed and Rewarded.
3. Sad Path: The video is Showed but dismissed.
4. Sad Path: The video Failed to show.

Since we can not resolve the promise 2 times we can not handle the both paths and both errors will be `any`. One solution for it will be to always return `call.resolve()` in the native part with an object like `{error?: AdMobError, value: sometype}`.

To be honest I am not fan of Promises for reasons like the problem above and maintain both promises and events is over complicated.

How about still using `show(): Promise<void>` and handle everything with `addEventListener` ? It is type safe and easier to find an error and to test!

distante commented:
I have let the happy paths on the plugin promise calls.

`Interstitial` is not triggering the correct events, also banner had wrong events and typescript definitions. It is now fixed but I have to eventually write some unit tests there.

rdlabo commented:
@distante  Thanks for great work!!

> How about still using show(): Promise<void>

I am troubled. I think EventListener is difficult for beginner, and sometimes forget remove EventListener. So I'm happy that this update will allow for a minimal implementation without event listeners.

```
Happy Path: The video is Showed.
Happy Path: The video is Showed and Rewarded.
Sad Path: The video is Showed but dismissed.
Sad Path: The video Failed to show.
```

I suggest `show()` get 2 stories.

```
Happy Path: User got Rewarded.
Sad Path: User didn't got Rewarded.
```

And user can get more information from event listeners. How about this??

distante commented:
Hi @rdlabo i have done it in the way you said. I just resolve the call with the reward for the show call or an error if that happened while the request is done.

The other parts as you said, are just available thru event listeners. 😁

I hope I can finish the Interstitial someday this week. But the Typescript definitions for Banner and Reward are done.

You can see in the angular example how all the reward and banner events are triggered . 🚀

rdlabo commented:
Great!! I will check soon!
and I think I will implement new demo code recent day!!

rdlabo commented:
@distante Hi. I'm working to follow new event on iOS.

I notice event that do not exist on iOS is added :
https://github.com/capacitor-community/admob/blob/feat/v3/src/banner/banner-definitions.interface.ts#L111-L114

iOS can't get receive bannerAdClicked, so I suggest this is be remove and instead to use "bannerAdOpened". How do you think?

iOS Event: https://github.com/googleads/googleads-mobile-ios-examples/blob/master/Swift/advanced/APIDemo/APIDemo/AdMobAdDelegateViewController.swift

distante commented:
Yes I suppose we can delete `Clicked` from the definitions @rdlabo.

I am not what the difference between `Clicked` and `Open` is. I think `Clicked` can also open stuff externally?

Anyway we can just delete it from the `.ts` file :) 

BTW, check that in the iOS part you define `Interstitial` events and not `RewardedInterstitial`. `RewardedInterstitial` is still in beta.

Or do we want to support both?

rdlabo commented:
> difference between Clicked and Open is.

Looks almost the same. It's the difference between click event or display event. The click event is an Android-only option, so remove it👍

iOS's `Rewarded Interstitial` is be released, but Android...🤔

Yes, I will remove

rdlabo commented:
> Or do we want to support both?

It's a difficult problem. These expect the user to behave in much the same way. I think `RewardedInterstitial` only is good(will implements in iOS).
@distante  Do you think??

distante commented:
I can implemented on Android also, but we have to keep both :

`Interstitial` and `RewardedInterstitial`. Both have the same events but `RewardedInterstitial` also has a `Rewarded`

rdlabo commented:
Oh. ...Yes, another Adsense format...

![スクリーンショット 2021-04-23 15 29 23](https://user-images.githubusercontent.com/9690024/115828444-b10dbc00-a448-11eb-860c-831358c2cc22.png)

Let's implement both in the future(T_T)

rdlabo commented:
I don't think it's necessary to implement RewardedInterstitial now. Work by future's us.

distante commented:
Ok let us just keep `Interstitial` for now, That is why in the Android code I split them into folders

```
./banner
./reward
./interstitial
```

It is easy later to just add a new one or delete an deprecated one :)

rdlabo commented:
> It is easy later to just add a new one or delete an deprecated one :)

YES, nice awesome work!!

App Open Ads is too: https://developers.google.com/admob/android/app-open-ads 
But Google said it is evil: https://techcrunch.com/2015/09/01/death-to-app-install-interstitials/#.iziele:SjwN
So I'm not very enthusiastic about this.

rdlabo commented:
In iOS, Interstitial and RewardVideo event is removed, and fullscreen event was added in its place.
https://developers.google.com/admob/ios/migration

These event remains on Android, but it seems better to use fullscreen event.

https://github.com/capacitor-community/admob/blob/feat/v3/src/interstitial/interstitial-ad-plugin-events.enum.ts#L7-L19
https://github.com/capacitor-community/admob/blob/feat/v3/src/reward/reward-ad-plugin-events.enum.ts#L7-L26

@distante How do you think??

distante commented:
The interstitial part is not complete done by my eh!  I just saw. there is an error here: `Loaded = 'onRewardedVideoAdLoaded',`
 it should be onInterstitialAdLoaded', I pushed a fix.

> In iOS, Interstitial and RewardVideo event is removed, and fullscreen event was added in its place.
> https://developers.google.com/admob/ios/migration
> 
> These event remains on Android, but it seems better to use fullscreen event.
> 
> https://github.com/capacitor-community/admob/blob/feat/v3/src/interstitial/interstitial-ad-plugin-events.enum.ts#L7-L19
> https://github.com/capacitor-community/admob/blob/feat/v3/src/reward/reward-ad-plugin-events.enum.ts#L7-L26
There is still an `OnUserEarnedRewardListener` for android https://developers.google.com/admob/android/migration

And it is what is used for `RewardAdPluginEvents.Rewarded` https://github.com/capacitor-community/admob/blob/e145d2235f2f464866e10bc6836e1a88607fbc5a/android/src/main/java/com/getcapacitor/community/admob/rewarded/RewardedAdCallbackAndListeners.kt#L17-L23

The equivalent iOS part is `GADUserDidEarnRewardHandler` 

```
func showRewardedAd() {
  ...
  if let ad = rewardedAd {
      ad.present(fromRootViewController: self,
               userDidEarnRewardHandler: {
                 let reward = ad.adReward
                 // TODO: Reward the user.
               }
      )
  } else {
    print("Ad wasn't ready")
  }
}
```

Fullscreen events are just to let the plugin consumer known the ad is being showed or dismissed. We should keep it I think.

> And maybe [BannerAdSize](https://github.com/capacitor-community/admob/blob/feat/v3/android/src/main/java/com/getcapacitor/community/admob/banner/BannerAdSizeEnum.java#L13) should be removed: https://developers.google.com/admob/android/banner#banner_sizes
> 

The `BannerAdSize` is mostly to let the user choose a size more easily: 

Instead of 
```
{
   adSize: 'FULL_BANNER'
}
```

```
{
   adSize: BannerAdSize.FULL_BANNER
}
```

and in the Java enum we can map the user option to whatever we need type safely like
```
        if(config.size !== BannerAdSizeEnum.SMART) {
            /// do the regular thing
        } else {
            /// DO special SMART thing
        }
```

rdlabo commented:
> Loaded = 'onRewardedVideoAdLoaded', it should be onInterstitialAdLoaded', I pushed a fix.

Yes, I will follow!!👍

> The equivalent iOS part is GADUserDidEarnRewardHandler

`Loaded`, `FailedToLoad` can be implements.  But `Showed` ,  `FailedToShow` , `Dismissed` must be called by delegate of full-screen. And there is no distinction between Interstitial and RewardedVideo.

From an iOS perspective, we should has common event of Full-screen API(Interstitial, RewardedVideo). This is simple.

```ts
export enum FullScreenEvents {
  // Each Platform implements
  Loaded = 'onFullScreenContentLoaded',
  FailedToLoad = 'onFullScreenContentFailedToLoad',
  Rewarded = 'onFullScreenContentRewarded',

  // Library of FullScreen API
  adDidPresentFullScreenContent = 'adDidPresentFullScreenContent',
  didFailToPresentFullScreenContentWithError = 'didFailToPresentFullScreenContentWithError',
  adDidDismissFullScreenContent= 'adDidDismissFullScreenContent',
}
```

If we use each name, in iOS implements is here. this ignites on same time, and I don't hope.

```
public func adDidPresentFullScreenContent(_ ad: GADFullScreenPresentingAd) {
    self.notifyListeners("onRewardedVideoAdShowed", data: [:])
    self.notifyListeners("onInterstitialAdShowed", data: [:])
    self.notifyListeners("onInterstitialRewardedAdShowed", data: [:])
}
```

Thanks.

rdlabo commented:
> The BannerAdSize is mostly to let the user choose a size more easily:

Sorry, mistake type(T_T) It mean `FLUID` BannerAdSize is removed.

rdlabo commented:
> The equivalent iOS part is GADUserDidEarnRewardHandler

`Loaded`, `FailedToLoad` can be implements.  But `Showed` ,  `FailedToShow` , `Dismissed` must be called by delegate of full-screen. And there is no distinction between Interstitial and RewardedVideo.

```
// All occur at the same time!
public func adDidPresentFullScreenContent(_ ad: GADFullScreenPresentingAd) {
    self.notifyListeners("onRewardedVideoAdShowed", data: [:])
    self.notifyListeners("onInterstitialAdShowed", data: [:])
    self.notifyListeners("onInterstitialRewardedAdShowed", data: [:])
}
```

distante commented:
> > The equivalent iOS part is GADUserDidEarnRewardHandler
> 
> `Loaded`, `FailedToLoad` can be implements.  But `Showed` ,  `FailedToShow` , `Dismissed` must be called by delegate of full-screen. And there is no distinction between Interstitial and RewardedVideo.
> 
> ```
> // All occur at the same time!
> public func adDidPresentFullScreenContent(_ ad: GADFullScreenPresentingAd) {
>     self.notifyListeners("onRewardedVideoAdShowed", data: [:])
>     self.notifyListeners("onInterstitialAdShowed", data: [:])
>     self.notifyListeners("onInterstitialRewardedAdShowed", data: [:])
> }
> ```

Ah yes the events are the same they both are `GADFullScreenPresentingAd` on iOS or `FullscreenContentCallback` on Android. But since the trigger is different I have a different instance for each on Android.

It helps the user react differently for Interstitial or reward `show` events, plus we can catch problems that comes from one or the other easily 😊

distante commented:
> > > The equivalent iOS part is GADUserDidEarnRewardHandler
> > 
> > 
> > `Loaded`, `FailedToLoad` can be implements.  But `Showed` ,  `FailedToShow` , `Dismissed` must be called by delegate of full-screen. And there is no distinction between Interstitial and RewardedVideo.
> > ```
> > // All occur at the same time!
> > public func adDidPresentFullScreenContent(_ ad: GADFullScreenPresentingAd) {
> >     self.notifyListeners("onRewardedVideoAdShowed", data: [:])
> >     self.notifyListeners("onInterstitialAdShowed", data: [:])
> >     self.notifyListeners("onInterstitialRewardedAdShowed", data: [:])
> > }
> > ```
> 
> Ah yes the events are the same they both are `GADFullScreenPresentingAd` on iOS or `FullscreenContentCallback` on Android. But since the trigger is different I have a different instance for each on Android.
> 
> It helps the user react differently for Interstitial or reward `show` events, plus we can catch problems that comes from one or the other easily 😊

I though I little bit more about this and I think I have a better idea for `FullscreenContentCallback` code in android.

I will create a base enum that has the full screen events `Showed`, `FailedToShow`, and `Dismissed`. And `InterstitialAdPluginEvents` and `RewardAdPluginEvents` will implement them, so we can use just one class or function that takes an enum that implements the base enum as parameter.

In that way the code is just one, tested just one time and no duplicate code will exist! 😃

distante commented:
@rdlabo take a look at this, I have created a subclass that implements the full screen callback, that takes an object like enum for the events:

https://github.com/capacitor-community/admob/blob/0f408f0363128c897fcb6573c80209d3c560c7ba/android/src/main/java/com/getcapacitor/community/admob/helpers/FullscreenPluginCallback.kt#L10-L27

I had to change the interstitial and reward plugin events enums to `objects` that implement an interface for it but now the code is not duplicated and we can use them for other things like `Reward Interstitial` or any other ads that needs full screen callback 😄 

```

interface LoadPluginEventNames {
    val Showed: String
    val FailedToShow: String
    val Dismissed: String
}

object InterstitialAdPluginPluginEvent: LoadPluginEventNames {
    const val Loaded = "interstitialAdLoaded"
    const val FailedToLoad = "interstitialAdFailedToLoad"
    override val Showed = "interstitialAdShowed"
    override val FailedToShow = "interstitialAdFailedToShow"
    override val Dismissed = "interstitialAdDismissed"
}

```


I also removed `FLUID` from the banners but I also added `SMART_BANNER` since it is still a supported banner size.

Take a look at the Angular Demo, All events are registered there.

I still have to check a couple of test because I moved some logic, but I think all the Android functional parts are now done.

rdlabo commented:
> I also removed FLUID from the banners but I also added SMART_BANNER since it is still a supported banner size.

Thanks!! 

Should we support to deprecated API? Smart Banner is deprecated in iOS.

![スクリーンショット 2021-04-26 15 33 59](https://user-images.githubusercontent.com/9690024/116038906-08a86380-a6a5-11eb-8de4-dc174cb4a2c3.png)

We can still use the Smart Banner, but it seems better to clarify that it is deprecated.

>  full screen callback

Let's take a closer look. Thank you for the information!

distante commented:
> > I also removed FLUID from the banners but I also added SMART_BANNER since it is still a supported banner size.
> 
> Thanks!!
> 
> Should we support to deprecated API? Smart Banner is deprecated in iOS.
> 
> ![スクリーンショット 2021-04-26 15 33 59](https://user-images.githubusercontent.com/9690024/116038906-08a86380-a6a5-11eb-8de4-dc174cb4a2c3.png)
> 
> https://developers.google.com/admob/ios/banner/smart
> 
> We can still use the Smart Banner, but it seems better to clarify that it is deprecated.
> 

Oh man! The google developers should communicate more one with the other. It is not mark as deprecated for Android:
![image](https://user-images.githubusercontent.com/5490201/116054054-97a08600-a67b-11eb-9575-bd400e477bcf.png)
https://developers.google.com/admob/android/banner/smart

Yeah I suppose we should mark it as deprecated first. It is not so hard to remove it after I think...

distante commented:
> Today, I update Angular demo to be able to check some expect api:
> https://twitter.com/rdlabo/status/1386678109768880128

That is really nice!

rdlabo commented:
@distante I did change some android code: https://github.com/capacitor-community/admob/pull/68/commits/0e15f82c6f226c679a63f7a83ef4468e542b7d8b

- prepare method is changed to do resolve
- "reason" in error message changed to "message"

Please check them🙏

rdlabo commented:
In my environment, I can't get height value:
https://github.com/capacitor-community/admob/blob/feat/v3/android/src/main/java/com/getcapacitor/community/admob/banner/BannerExecutor.java#L230

ex: `{"width":411,"height":0}`

@distante how about in your environment??

distante commented:
> @distante I did change some android code.
> 
> 1. [9ba0bc2](https://github.com/capacitor-community/admob/commit/9ba0bc2f7daec7bef4d14a3773db4a696fa1e7be)
> 
> * at resume, banner size get non zero size
> * loaded event get void
> * at failed, mAdView destroy
> 
> 1. [0e15f82](https://github.com/capacitor-community/admob/commit/0e15f82c6f226c679a63f7a83ef4468e542b7d8b)
> 
> * prepare method is changed to do resolve
> * "reason" in error message changed to "message"
> 
> Please check them🙏

Looks good to me



> In my environment, I can't get height value at onAdLoaded:
> https://github.com/capacitor-community/admob/blob/feat/v3/android/src/main/java/com/getcapacitor/community/admob/banner/BannerExecutor.java#L230
> 
> ex: `{"width":411,"height":0}`
> 
> @distante how about in your environment??

Yeah, my bad, there was an error in the [BannerAdSize](https://github.com/capacitor-community/admob/pull/68/commits/44e5acec331b7e6d0dbad73aa4f2d495b27819a7) class, I fix it. I also removed the server config in `capacitor.config` because it does not work with my IP range.

---

BTW I do not see the new layout in the demo app:
![image](https://user-images.githubusercontent.com/5490201/116191346-6df36780-a72c-11eb-9840-f9e8b74ec1d2.png)

rdlabo commented:
> BTW I do not see the new layout in the demo app:

@distante did you run `npm run cap` in demo folder?

distante commented:
> > BTW I do not see the new layout in the demo app:
> 
> @distante did you run `npm run cap` in demo folder?

I runned `ionic build && npx cap sync`

rdlabo commented:
mumu🤔 Please try `ionic serve`. I changed demo to tabs layout now.

distante commented:
Will do after work :)

distante commented:
> > > BTW I do not see the new layout in the demo app:
> > 
> > 
> > @distante did you run `npm run cap` in demo folder?
> 
> I runned `ionic build && npx cap sync`

soooo!! the problem was that I used `ionic build` but I do not have ionic as global anymore! I needed to run `ng build` 
---
You know what? It would be really cool if we could run some e2e test on device in the demo project, in that way we could catch really fast some regression that the unit test did not catch. 

It is something that lacks documentation though 🤔  https://ionicframework.com/integrations/appium

rdlabo commented:
> there was an error in the BannerAdSize class, I fix it

Thanks for work. demo is changed to be all green!!

<img width="1291" alt="スクリーンショット 2021-04-28 9 06 38" src="https://user-images.githubusercontent.com/9690024/116327159-1541ce80-a801-11eb-9ef0-ecaaafc9af27.png">

> It is something that lacks documentation though

No, I've never heard of it before. But looks good, and this seems to be a high affinity with this project. I'll do some reading.

rdlabo commented:
@distante I changed some code for ci test:
https://github.com/capacitor-community/admob/pull/68/commits/dac20aadb887ec705417aa498602f755914f091b

Every change is updating library version, or auto formatter changes. The scope of the change is large, so I will contact you.

rdlabo commented:
@distante [@]capacitor/docgen is not support extends of typescript. Do you have any better ideas?

distante commented:
> @distante I changed some code for ci test:
> [dac20aa](https://github.com/capacitor-community/admob/commit/dac20aadb887ec705417aa498602f755914f091b)
> 
> Every change is updating library version, or auto formatter changes. The scope of the change is large, so I will contact you.

Wow that looks complicated! I never set up an Native CI, I just did a couple for node projects.

I have added the required config to run the unit tests, As you see, the missing test are now correctly red 👍  (or 👎  ??)

I will try to fix them after work.

rdlabo commented:
@distante Yes, thanks! The Android test failed correctly now.
And I added ios test(test code is empty, only environment settings).

distante commented:
> capacitor/docgen

Maybe we should open a request for this here https://github.com/ionic-team/capacitor-docgen/issues ? Maybe it gets fixed before Capacitor 3 is released.

rdlabo commented:
> Maybe we should open a request

@distante I done. and forked and created patch. Right now, we can check API at README.
https://github.com/capacitor-community/admob/tree/feat/v3#index

distante commented:
@rdlabo I fixed the tests and added some missing ones. All checks pass now.

> > Maybe we should open a request
> 
> @distante I done. and forked and created patch. Right now, we can check API at README.
> https://github.com/capacitor-community/admob/tree/feat/v3#index
> 
> If you run `npm run build` or `npm run docgen`, please check here: https://github.com/capacitor-community/admob/blob/feat/v3/CONTRIBUTING.md#temporary-docgen

we can also change `@capacitor/docgen` in package.json to point to your repo/branch until it gets merged, in that way we do not need nothing extra.

rdlabo commented:
> I fixed the tests and added some missing ones. All checks pass now.

Thanks for the quick work!

> package.json to point to your repo/branch

To do so, need to change the `.gitignore` and commit the build results. I'll wait for the Ionic team to respond in a bit, and decide whether to publish in a separate package or not.

rdlabo commented:
@distante That's what I said, but let's just get it directly from GitHub!

rdlabo commented:
@distante If we follow the events of `InterstitialAdPluginEvents` and `RewardAdPluginEvents`, then `BannerAdPluginEvents.Loaded` should be `BannerAdPluginEvents.Showed`. What do you think?

- BannerAdPluginEvents.Loaded: () => void

- InterstitialAdPluginEvents.Loaded: (info) => void
- InterstitialAdPluginEvents.Showed: () => void

- RewardAdPluginEvents.Loaded: (info) => void
- RewardAdPluginEvents.Showed: () => void

distante commented:
> @distante If we follow the events of `InterstitialAdPluginEvents` and `RewardAdPluginEvents`, then `BannerAdPluginEvents.Loaded` should be `BannerAdPluginEvents.Showed`. What do you think?
> 
> * BannerAdPluginEvents.Loaded: () => void           // Showed is not exist
> * InterstitialAdPluginEvents.Loaded: (info) => void
> * InterstitialAdPluginEvents.Showed: () => void
> * RewardAdPluginEvents.Loaded: (info) => void
> * RewardAdPluginEvents.Showed: () => void

mmm I am not sure, for one side `show` does make sense, but the native event is a load event. Also it triggers automatically each time the banner changes without being called.

rdlabo commented:
> mmm I am not sure

So let's keep going with this. Thank you for your consideration!

rdlabo commented:
Todo: 

- [BannerAdOptions](https://github.com/capacitor-community/admob/tree/feat/v3#banneradoptions) is not implements AdOptions
- Check comment at interfaces

rdlabo commented:
> BannerAdOptions is not implements AdOptions

It is difficult to get only the extends, so we commented it out.

rdlabo commented:
It looks like it's ready for release, and we'll merge this branch in when we officially release Capacitor3.

distante commented:
@rdlabo I added a control handling for `show` calls (interstitial and reward) for failed or non loaded ads. If the user forgot to prepare the ads the android app crashed (see #102)

distante commented:
Oh.. I just did it for the Interstitial... I will try to check the reward tomorrow.
